R/github-pat.R

```diff
@@ -66,6 +69,36 @@ set_github_pat <- function(force_new = FALSE, validate = interactive(), verbose
   return(FALSE)
 }
 
+# Some notes:
+# - Format: GITHUB_PAT_({user}_AT_){host}
+# - By default, tokens get stored as user 'PersonalAccessToken' in the credential
+#   store, to distinguish them from user/password credentials that the user may have
+#   stored. You can override this via the URL, but better not.
+# - GitHub usernames (incl GHE) can only contain alphanumeric characters and dashes.
+# - Hostnames are case insensitive. We normalize hostname tolower, so you can
+#   safely parse _AT_ to look for custom usernames.
+get_pat_credential_url <- function(url){
+  # strip protocol
+  url <- sub("^[a-z]+://", "", tolower(url))
+  # strip path
+  url <- sub("/.*$", "", url)
+  url <- sub("api.github.com", "github.com", url, fixed = TRUE)
+  pat_url <- if(grepl("@", url, fixed = TRUE)){
+    paste0('https://', url)
+  } else {
+    paste0('https://PersonalAccessToken@', url)
+  }
+  pat_var <- if(grepl("github.com$", url)){
+    'GITHUB_PAT'
+  } else {
+    paste0('GITHUB_PAT_', gsub("\\.", "_", sub("^.+@", "", url)))
```

Comment:
gh does `toupper(url)` (modulo the fact that it does not have any notion of such a URL containing a username).